openmls/src/test_utils/test_framework/mod.rs

```diff
@@ -322,18 +336,54 @@ impl MlsGroupTestSetup {
             .expect("An unexpected error occurred.");
         let sender_groups = sender.groups.read().expect("An unexpected error occurred.");
         let sender_group = sender_groups
-            .get(&group.group_id)
+            .get(group.group_id())
             .expect("An unexpected error occurred.");
-        group.members = sender
-            .get_members_of_group(&group.group_id)?
-            .iter()
-            .map(
-                |Member {
-                     index, credential, ..
-                 }| { (index.usize(), credential.identity().to_vec()) },
-            )
-            .collect();
-        group.public_tree = sender_group.export_ratchet_tree();
+
+        let signature_pk = sender_group.own_leaf().unwrap().signature_key();
+        let signer = SignatureKeyPair::read(
+            sender.crypto.key_store(),
+            signature_pk.as_slice(),
+            sender_group.ciphersuite().signature_algorithm(),
+        )
+        .unwrap();
+
+        // Update the public group. If it's a public message, just process it. Create a new group from the sender.
+        match message {
+            ProtocolMessage::PrivateMessage(_) => {
+                // If it's a ciphertext we can't track group membership with a
+                // PublicGroup, so we have to create a new one every time with
+                // the sender's state.
+                let (new_public_group, _extensions) = PublicGroup::from_external(
```

Comment:
As noted offline, creating a group will involve validating all signatures going forward and this step will become really expensive. Maybe add a comment at least.